File: server/repositories/chatRepo.mjs
import { readSeedJson } from '../utils/seedLoader.mjs';
import { runOracleQuery } from '../db/oracleClient.mjs';
import { getServerConfig } from '../config/env.mjs';
import { createOracleUnavailableError } from '../errors/oracleErrors.mjs';
import { classifyIntent, getChatIntentCatalog, listIntentSuggestions } from './chatIntentsRepo.mjs';
import { buildChatContext } from '../services/chatContextService.mjs';

function nowTime() {
  return new Date().toLocaleTimeString('pt-BR', { hour: '2-digit', minute: '2-digit' });
}

function normalizeText(value) {
  return String(value || '')
    .toLowerCase()
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '');
}

export async function getChatBootstrap(condominiumId = 1) {
  const { dbDialect, allowOracleSeedFallback } = getServerConfig();
  const catalog = getChatIntentCatalog();

  if (dbDialect === 'oracle') {
    try {
      const [overdueRows, alertsRows] = await Promise.all([
        runOracleQuery(`
          select count(1) as TOTAL
          from mart.vw_financial_invoices
          where lower(status) = 'overdue'
            and condominio_id = :condominiumId
        `, { condominiumId }),
        runOracleQuery(`
          select count(1) as TOTAL
          from mart.vw_alerts_operational
          where lower(gravidade) in ('alta', 'critica')
            and condominio_id = :condominiumId
        `, { condominiumId }),
      ]);

      const overdueCount = Number(overdueRows?.[0]?.TOTAL || 0);
      const criticalAlertsCount = Number(alertsRows?.[0]?.TOTAL || 0);

      return {
        welcomeMessage: `Contexto Oracle carregado: ${overdueCount} faturas vencidas e ${criticalAlertsCount} alertas criticos.`,
        catalogVersion: catalog.version,
        suggestions: listIntentSuggestions(3),
      };
    } catch (error) {
      if (!allowOracleSeedFallback) {
        throw createOracleUnavailableError(error);
      }
    }
  }

  const seed = readSeedJson('chat_bootstrap.json');
  return {
    ...seed,
    catalogVersion: catalog.version,
    suggestions: listIntentSuggestions(3),
  };
}

function buildIntentResponse(intentId, context) {
  if (intentId === 'financial_priorities') {
    return `Prioridade financeira: ${context.metrics.overdueInvoices} faturas vencidas e ${context.metrics.pendingInvoices} pendentes. Recomendo acao imediata nas vencidas e trilha de cobranca para pendentes.`;
  }
  if (intentId === 'critical_alerts') {
    return `Alertas criticos ativos: ${context.metrics.criticalAlerts}. Recomendo validar causa raiz, acionar manutencao e registrar responsavel por cada evento.`;
  }
  if (intentId === 'action_plan') {
    return `Plano de acao 24h: (1) tratar ${context.metrics.criticalAlerts} alertas criticos, (2) reduzir ${context.metrics.overdueInvoices} vencidas, (3) revisar ${context.metrics.maintenanceUnits} unidades em manutencao.`;
  }
  if (intentId === 'consumption_anomalies') {
    return `Visao de consumo: ${context.metrics.criticalAlerts} eventos criticos podem impactar telemetria. Priorize unidades em manutencao (${context.metrics.maintenanceUnits}) para reduzir risco de anomalia.`;
  }

  return `Resumo atual: ${context.metrics.overdueInvoices} vencidas, ${context.metrics.pendingInvoices} pendentes, ${context.metrics.criticalAlerts} alertas criticos e ${context.metrics.maintenanceUnits} unidades em manutencao.`;
}

function evaluateGuardrails(message, classification) {
  const normalized = normalizeText(message);
  const outOfScopeTerms = ['futebol', 'jogo', 'receita', 'culinaria', 'filme', 'serie', 'celebridade', 'astrologia', 'loteria', 'piada'];
  const outOfScopeDetected = outOfScopeTerms.some((term) => normalized.includes(term));

  if (outOfScopeDetected) {
    return {
      blocked: true,
      reason: 'OUT_OF_SCOPE',
      policyVersion: 's5-03.v1',
      confidence: classification.confidence,
      message: 'Pergunta fora do escopo operacional do condominio.',
    };
  }

  if (classification.confidence === 'low') {
    return {
      blocked: true,
      reason: 'LOW_CONFIDENCE',
      policyVersion: 's5-03.v1',
      confidence: classification.confidence,
      message: 'Confianca insuficiente para responder com seguranca. Reformule com contexto de alertas, consumo, faturas ou gestao.',
    };
  }

  return {
    blocked: false,
    reason: null,
    policyVersion: 's5-03.v1',
    confidence: classification.confidence,
    message: null,
  };
}

export async function askChat(message, condominiumId = 1) {
  const classification = classifyIntent(message);
  const context = await buildChatContext(condominiumId);
  const guardrails = evaluateGuardrails(message, classification);
  const text = guardrails.blocked
    ? guardrails.message
    : buildIntentResponse(classification.intentId, context);

  return {
    id: `bot-${Date.now()}`,
    role: 'assistant',
    text,
    time: nowTime(),
    intentId: classification.intentId,
    confidence: classification.confidence,
    promptCatalogVersion: classification.catalogVersion,
    sources: context.sources,
    limitations: 'Resposta automatica com base no contexto operacional atual; confirme casos criticos com responsavel tecnico.',
    guardrails: {
      blocked: guardrails.blocked,
      reason: guardrails.reason,
      policyVersion: guardrails.policyVersion,
    },
  };
}

File: backend/app/repositories/chat_repo.py
"""
Chat repository — thin adapter over the LangGraph multi-agent pipeline.

ask_chat() delegates all AI logic to the StateGraph defined in app.ai.graph.
get_chat_bootstrap() remains a standalone DB query (no graph involvement).
"""
from __future__ import annotations

import asyncio
import logging

from app.core.config import settings
from app.core.errors import create_oracle_unavailable_error
from app.core.tenancy import ensure_condominium_id
from app.db.oracle_client import run_oracle_query
from app.observability.metrics_store import record_api_fallback_metric
from app.repositories.chat_intents_repo import get_chat_intent_catalog, list_intent_suggestions
from app.utils.seed_loader import read_seed_json
from app.ai.memory import get_memory, save_to_memory
from app.ai.graph import get_agent_graph

_log = logging.getLogger(__name__)


async def get_chat_bootstrap(condominium_id: int) -> dict:
    condominium_id = ensure_condominium_id(condominium_id)
    catalog = get_chat_intent_catalog()
    ai_badge = "Gemini AI" if settings.gemini_api_key else "Modo regras"

    if settings.db_dialect == "oracle":
        try:
            overdue_rows, alerts_rows = await asyncio.gather(
                run_oracle_query(
                    """
                    select count(1) as TOTAL
                    from mart.vw_financial_invoices
                    where lower(status) = 'overdue'
                      and condominio_id = :condominiumId
                    """,
                    {"condominiumId": condominium_id},
                ),
                run_oracle_query(
                    """
                    select count(1) as TOTAL
                    from mart.vw_alerts_operational
                    where lower(gravidade) in ('alta', 'critica')
                      and condominio_id = :condominiumId
                    """,
                    {"condominiumId": condominium_id},
                ),
            )
            overdue_count = int((overdue_rows or [{}])[0].get("TOTAL") or 0)
            critical_count = int((alerts_rows or [{}])[0].get("TOTAL") or 0)
            welcome = (
                f"Ola! Sou o **CondoGuard Copiloto** ({ai_badge}). "
                f"Situacao atual: **{overdue_count} faturas vencidas** e **{critical_count} alertas criticos**. "
                "Pergunte sobre faturas, alertas, consumo, contratos ou gestao de unidades."
            )
            return {
                "welcomeMessage": welcome,
                "catalogVersion": catalog["version"],
                "suggestions": list_intent_suggestions(3),
            }
        except Exception as exc:
            if not settings.allow_oracle_seed_fallback:
                raise create_oracle_unavailable_error(exc)
            record_api_fallback_metric("chat", "oracle_fallback_seed")

    seed = read_seed_json("chat_bootstrap.json")
    welcome = (
        f"Ola! Sou o **CondoGuard Copiloto** ({ai_badge}). "
        "Posso ajudar com faturas, alertas operacionais, consumo de recursos e gestao de unidades. "
        "Como posso ajudar hoje?"
    )
    return {
        **seed,
        "welcomeMessage": welcome,
        "catalogVersion": catalog["version"],
        "suggestions": list_intent_suggestions(3),
    }


async def ask_chat(
    message: str,
    condominium_id: int,
    session_id: str | None = None,
) -> dict:
    condominium_id = ensure_condominium_id(condominium_id)
    memory = get_memory(session_id)
    history = memory.messages

    initial_state = {
        "session_id": session_id or "",
        "condominium_id": condominium_id,
        "question": message,
        "history": history,
        "context": {},
        "classification": {},
        "route": {},
        "guardrails": {"blocked": False, "reason": None, "policyVersion": "s5-03.v1", "message": None},
        "rag_docs": [],
        "agent_response": "",
        "agent_name": "",
        "ai_powered": False,
        "action_result": None,
        "collaboration": {},
        "final_response": {},
    }

    graph = get_agent_graph()
    result = await graph.ainvoke(initial_state)

    final = result.get("final_response", {})

    # Save exchange to session memory only when the LLM actually ran
    if session_id and final.get("aiPowered") and not result.get("guardrails", {}).get("blocked"):
        save_to_memory(memory, message, final.get("text", ""))

    return final
